internal/aws/awsutil/conn.go

```diff
@@ -369,17 +371,120 @@ func getSTSRegionalEndpoint(r string) string {
+}
+
 // getPartition return AWS Partition for the provided region.
 func getPartition(region string) string {
 	p, _ := endpoints.PartitionForRegion(endpoints.DefaultPartitions(), region)
 	return p.ID()
 }
+
+// order
+// 1. override creds providers
+// 2. shared creds file
+// 3. shared profile
+// explicit keys do not make sense in this context since contrib does not take in explicit keys
+// and cwa does not provide explicit keys to config but come from these methods of getting creds
+// @TODO make this more upstream friendly working in the future
+func getCredentialProviderChain(cfg *AWSSessionSettings) []credentials.Provider {
+	overrideCredProviders := override.GetCredentialsChainOverride().GetCredentialsChain()
+	var credProviders []credentials.Provider
+	for _, provider := range overrideCredProviders {
+		for _, file := range cfg.SharedCredentialsFile {
+			credProviders = append(credProviders, provider(file))
+		}
+	}
+	if cfg.Profile != "" || len(cfg.SharedCredentialsFile) > 0 {
+		if len(cfg.SharedCredentialsFile) == 0 {
+			credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{
+				sharedCredentialsProvider: &credentials.SharedCredentialsProvider{
+					Filename: "",
+					Profile:  cfg.Profile,
+				},
+			})
+		}
+		for _, file := range cfg.SharedCredentialsFile {
+			credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{
+				sharedCredentialsProvider: &credentials.SharedCredentialsProvider{
+					Filename: file,
+					Profile:  cfg.Profile,
+				},
+			})
+		}
+	}
+	return credProviders
+}
+
+func newStsCredentials(c client.ConfigProvider, roleARN string, region string) *credentials.Credentials {
+	regional := &stscreds.AssumeRoleProvider{
+		Client: sts.New(c, &aws.Config{
+			Region:              aws.String(region),
+			STSRegionalEndpoint: endpoints.RegionalSTSEndpoint,
+			HTTPClient:          &http.Client{Timeout: 1 * time.Minute},
+		}),
+		RoleARN:  roleARN,
+		Duration: stscreds.DefaultDuration,
+	}
+
+	fallbackRegion := getFallbackRegion(region)
+
+	partitional := &stscreds.AssumeRoleProvider{
+		Client: sts.New(c, &aws.Config{
+			Region:              aws.String(fallbackRegion),
+			Endpoint:            aws.String(getFallbackEndpoint(fallbackRegion)),
+			STSRegionalEndpoint: endpoints.RegionalSTSEndpoint,
+			HTTPClient:          &http.Client{Timeout: 1 * time.Minute},
+		}),
+		RoleARN:  roleARN,
+		Duration: stscreds.DefaultDuration,
+	}
+
+	return credentials.NewCredentials(&stsCredentialProvider{regional: regional, partitional: partitional})
+}
+
+func getFallbackRegion(region string) string {
+	partition := getEndpointPartition(region)
+	switch partition.ID() {
+	case bjsPartition:
+		return bjsFallbackRegion
+	case pdtPartition:
+		return pdtFallbackRegion
+	case dcaPartition:
+		return dcaFallbackRegion
+	case lckPartition:
+		return lckFallbackRegion
+	default:
+		return classicFallbackRegion
+	}
+}
+
+func getEndpointPartition(region string) endpoints.Partition {
```

Comment:
nit: `getPartition` can probably be removed and replaced with this one.

Reply:
I'm just going to keep the current contrib code for now and decide what to do later.